File: Algorithms/0576_Out_of_Boundary_Paths/Python/Out_of_Boundary_Paths_Solution_1.py

# Space: O(n)
# Time: O(n)


class Solution:

    def findPaths(self, m, n, maxMove, startRow, startColumn):
        self.mod = 1000000007
        self.board = [[[-1 for _ in range(n)] for _ in range(m)] for _ in range(maxMove + 1)]

        return self.dfs(m, n, startRow, startColumn, maxMove) % self.mod

    def dfs(self, row, column, y, x, n):
        if (not 0 <= y < row) or (not 0 <= x < column): return 1
        if (n == 0): return 0
        if self.board[n][y][x] > -1: return self.board[n][y][x]

        self.board[n][y][x] = (self.dfs(row, column, y - 1, x, n - 1)
                               + self.dfs(row, column, y, x - 1, n - 1)
                               + self.dfs(row, column, y + 1, x, n - 1)
                               + self.dfs(row, column, y, x + 1, n - 1))

        return self.board[n][y][x]

File: Algorithms/0576_Out_of_Boundary_Paths/Java/OutofBoundaryPathsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private int mod = 1000000007;
    private int[][][] board = null;

    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        board = new int[maxMove + 1][m][n];
        for (int[][] temp : board) {
            for (int[] row : temp) {
                Arrays.fill(row, -1);
            }
        }
        return (int)dfs(m, n, startRow, startColumn, maxMove);

    }
    private int dfs(int row, int column, int y, int x, int n) {
        if (!(0 <= y && y < row) || !(0 <= x && x < column)) return 1;
        if (n <= 0) return 0;
        if (board[n][y][x] != -1) return board[n][y][x];

        int tempRes = 0;
        tempRes = tempRes % mod + dfs(row, column, y - 1, x, n - 1) % mod;
        tempRes = tempRes % mod + dfs(row, column, y + 1, x, n - 1) % mod;
        tempRes = tempRes % mod + dfs(row, column, y, x - 1, n - 1) % mod;
        tempRes = tempRes % mod + dfs(row, column, y, x + 1, n - 1) % mod;

        board[n][y][x] = tempRes % mod;
        return board[n][y][x];
    }
}





